[develop] Updates to devclean.sh script and plotting scripts and tasks

natalie-perlin opened the pull request:
## DESCRIPTION OF CHANGES: 
-  ./devclean.sh script that cleans SRW builds is updated, all the cleaning tasks are done for the directories under the main SRW tree

- Documentation updated for the devclean.sh script changes

- plotting scripts updated to have geographical data visible over the colored fields

- plotting task updated to allow graphics output for individual ensemble members

- use python3 to checkout external sub-modules in a checkout_externals script; python3 is a default for other scripts; some systems such as MacOS no longer come with python2 

### Type of change
<!-- Please delete options that are not relevant. Add an X to check off a box. -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update (UPDATED)

## TESTS CONDUCTED: 

Tested various combinations of cleaning up tasks on MacOS

- [ ] hera.intel
- [ ] orion.intel
- [ ] hercules.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] derecho.intel
- [ ] gaea.intel
- [ ] gaeac5.intel
- [ ] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [ ] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

## DEPENDENCIES:

https://github.com/natalie-perlin/ufs-srweather-app/pull/12 (Merged)

## DOCUMENTATION:
<!-- If this PR is contributing new capabilities that need to be documented, please also include updates to the RST files (docs/UsersGuide/source) as supporting material. -->

## ISSUE: 
devclean.sh script update addresses the issue in #1073 (devclean.sh demonstrates very unsafe directory removal behavior)

## CHECKLIST
<!-- Add an X to check off a box. -->
- [x] My code follows the style guidelines in the Contributor's Guide
- [x] I have performed a self-review of my own code using the Code Reviewer's Guide
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [x] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## LABELS (optional): 
<!-- If you do not have permissions to add labels to your own PR, request that labels be added here. 
Add an X to check off a box. Delete any unnecessary labels. -->
A Code Manager needs to add the following labels to this PR: 
- [ ] Work In Progress
- [ ] bug
- [x] enhancement
- [x] documentation
- [ ] release
- [ ] high priority
- [ ] run_ci
- [ ] run_we2e_fundamental_tests
- [ ] run_we2e_comprehensive_tests
- [ ] Needs Cheyenne test 
- [ ] Needs Jet test 
- [ ] Needs Hera test 
- [ ] Needs Orion test 
- [ ] help wanted

## CONTRIBUTORS (optional): 
 @mkavulich

Commit:
plotting scripts updated to have geographical data visible over the colored fields

Commit:
plotting task updated to allow graphics output for individual ensemble members

Commit:
updating devclean.sh script

Commit:
use python3 in checkout_externals script

Commit:
Documentation update for the devclean.sh script

MichaelLueken commented on devclean.sh:
```diff
@@ -110,6 +108,9 @@ while :; do
   shift
 done
 
+# Make sure to be in the SRW main directory before any removeing or cleaning
```
removing

```suggestion
# Make sure to be in the SRW main directory before any removing or cleaning
```

MichaelLueken reviewed (commented):
@natalie-perlin -

I've noted a small grammatical issue in one of the `devclean.sh` comments.

Please make sure to remove the changes to `devclean.sh`, `parm/wflow_plot.yaml`, `scripts/exregional_plot_allvars.py`, and `scripts/exregional_plot_allvars_diff.py` from PR #1089.

You can use the following command to back out the changes to these files in the `rrfs_ics_lbcs` branch:

git checkout 28cbbc8 devclean.sh parm/wflow/plot.yaml scripts/exregional_plot_allvars.py scripts/exregional_plot_allvars_diff.py

This command will checkout the version of the four files at the `28cbbc8` hash of your branch, the hash before you began applying changes related to the PR.

natalie-perlin committed:
Update devclean.sh - typo

Co-authored-by: Michael Lueken <63728921+MichaelLueken@users.noreply.github.com>

MichaelLueken commented on devclean.sh:
```diff
@@ -4,30 +4,28 @@
       removes "bin", "build" directories, and other build artifacts
-  --remove
-      removes the "build" directory, keeps the "bin", "lib" and other build artifacts intact
+  --bin-dir=BIN_DIR
```
@mkavulich - Is this enough of a simplification, or are you wanting for @natalie-perlin to remove the `bin_dir`, `build_dir`, and `install_dir`, and have the `--remove` option only delete the build directory, the `-a` and `--clean` options to only remove the build, exec, include, lib, and share directories, the `--sub-modules` option to only remove the components from the sorc directory, and the `--conda` option to only remove the conda directory and conda_loc file?

natalie-perlin replied:
Note that we do need to keep $BIN_DIR specification, needed when working with software containers

MichaelLueken reviewed (approved):
@natalie-perlin -

Thanks for correcting the typo in `devclean.sh`!  As far as I'm concerned, I'm fine with these changes.  However, I'd like additional feedback from @mkavulich on the simplification of the script.  I will go ahead and approve this work now, but will hold off on final testing and merging until Mike has had the opportunity to check this update.

gsketefian committed:
[develop] Bug fix to support the %H format in METplus via printf. (#1102)

This bug was encountered when verifying forecast output that has a 2-digit forecast hour in its name. It turns out specifying the METplus format %H to obtain a 2-digit forecast hour in the workflow/verification configuration variable FCST_FN_TEMPLATE (and others) causes an error in the shell script eval_METplus_timestr_tmpl.sh because bash's printf utility does not support the %H format. This fixes that error using a similar approach to the %HHH format for obtaining 3-digit hours.

mkavulich committed:
Safety and simplification updates for devclean.sh

mkavulich commented:
@MichaelLueken @natalie-perlin Since I'm not properly communicating my desired changes to this script, I felt it best to just make those changes and propose them as a PR to this branch. Please review and offer any comments/questions: https://github.com/natalie-perlin/ufs-srweather-app/pull/12

mkavulich committed:
A couple more improvements:

 - Format, remove old comment
 - Only remove conda if the location in conda_loc is the same as the
default conda install

gspetro-NOAA committed:
[develop]: Update requests and certifi in requirements.txt (#1103)

* The Dependabot PR #1101 identified the need to update the certifi version, but requests should also be updated from the current (yanked) version in the requirements file.
* The README.md and doc/README files have also been updated.

mkavulich committed:
Changes based on initial round of comments:

 - Default case (no flags provided) now does nothing except print usage
and exit
 - Include "--build" flag to remove build artifacts
 - Include short aliases for all the removal options

mkavulich committed:
Hide another print behind --verbose flag

mkavulich committed:
Add --container flag for different bin directory in containers

christinaholtNOAA committed:
[develop] Transition the var_defns bash file to YAML. (#1098)

Use YAML for the configuration language at run time.

---------

Co-authored-by: Michael Lueken <63728921+MichaelLueken@users.noreply.github.com>
Co-authored-by: Michael Kavulich <kavulich@ucar.edu>
Co-authored-by: michael.lueken <Michael.Lueken@noaa.gov>

EdwardSnyder-NOAA committed:
[develop] Fixes for PW Jenkins Nightly Builds (#1091)

* Adds logic to handle GCP's default conda env, which conflicts with the SRW App's conda env. Fixes a Parallel Works naming convention bug in the script.
* It also addresses a known issue with a Ruby warning on PW instances that prevents the run_WE2E_tests.py from exiting gracefully. The solution we use in our bootstrap for /contrib doesn't seem to work for the /lustre directory, which is why the warning is hardcoded into the monitor_jobs.py script.
* The new spack-stack build on Azure is missing a gnu library, so added the path to this missing library to the proper run scripts and cleaned up the wflow noaacloud lua file.
* Removed log and error files from the qsub wrapper script so that qsub can generate these files with the job id in the files name. Also, fixed typo in the wrapper script.

EdwardSnyder-NOAA committed:
[develop] S3 doc updates (#1104)

As part of the data governance initiative, all s3 buckets will need some sort of versioning control. To meet these needs the AWS S3 Bucket was reorganized with the develop data stored under a 'develop-date' folder and the verification sample case and the document case (current_release_data) moved under a new folder called 'experiment-user-cases'.

---------

Co-authored-by: Michael Lueken <63728921+MichaelLueken@users.noreply.github.com>

MichaelLueken committed:
[develop] Update ufs-weather-model hash and further clean the machines tested in PULL_REQUEST_TEMPLATE (#1096)

* Update ufs-weather-model hash to b5a1976 (July 30)
* Add hera.gnu, remove cheyenne.intel, cheyenne.gnu, and gaeac5.intel, and alphabetize the machines in the TESTS CONDUCTED section of the PULL_REQUEST_TEMPLATE
* Correct behavior of Jenkins Functional WorkflowTaskTests. Currently, TASK_DEPTH is set to null, resulting in no tests being run during the Functional WorkflowTaskTests stage. Replaced env with params in Jenkinsfile for setting TASK_DEPTH. Testing shows that this will correctly set TASK_DEPTH to the default value of 9 and allow the tests to run
* Removed extraneous entries from the verification scripts to remove KeyError messages in the associated verification log files
* Reapplied necessary modification to modulefiles/tasks/noaacloud/plot_allvars.local.lua to allow plotting tasks to run on NOAA cloud platforms

natalie-perlin committed:
Update devclean.sh

Updates and simplifications of the devclean.sh script

natalie-perlin committed:
Merge pull request #12 from mkavulich/scripts_plots_updates

Safety and simplification updates for devclean.sh

natalie-perlin committed:
Merge branch 'develop' into scripts_plots_updates

natalie-perlin commented:
All pull request by @mkavulich has been merged https://github.com/natalie-perlin/ufs-srweather-app/pull/12, and all the questions resolved.

Commit:
Updating a FAQ.rst doc file, and updating help section of the devclean.sh script

MichaelLueken commented on parm/wflow/plot.yaml:
```diff
@@ -24,25 +26,30 @@ default_task_plot: &default_task
+      cycledef: '{% for h in range(0, workflow.LONG_FCST_LEN+1) %}{% if h <= workflow.FCST_LEN_CYCL|min %}forecast {% else %}long_forecast {% endif %}{% endfor %}'
+    task_plot_allvars_mem#mem#_f#fhr#:
+      <<: *default_task
+      command: '&LOAD_MODULES_RUN_TASK_FP; "plot_allvars" "&JOBSdir;/JREGIONAL_PLOT_ALLVARS"'
```
Following the merge of PR #1098, `LOAD_MODULES_RUN_TASK_FP` must be changed to `LOAD_MODULES_RUN_TASK`, otherwise, the plotting scripts will fail.

```suggestion
      command: '&LOAD_MODULES_RUN_TASK; "plot_allvars" "&JOBSdir;/JREGIONAL_PLOT_ALLVARS"'
```

MichaelLueken reviewed (changes requested):
@natalie-perlin -

While running the fundamental WE2E tests on Hera, the `grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v17_p8_plot` test failed.  In PR #1098, the default `LOAD_MODULES_RUN_TASK_FP` was changed to `LOAD_MODULES_RUN_TASK`.  When this change was applied to `parm/wflow/plot.yaml` in your `scripts_plots_updates` branch, the test successfully ran.  Please make this modification in your `parm/wflow/plot.yaml` file.

Additionally, please convert one of the currently existing ensemble WE2E tests to exercise the plotting capability that is being added in this PR.  Adding a new feature without adding the ability to test it is bad practice.

natalie-perlin committed:
Update parm/wflow/plot.yaml, LOAD_MODULES_RUN_TASK command

Co-authored-by: Michael Lueken <63728921+MichaelLueken@users.noreply.github.com>

natalie-perlin committed:
update devclean.sh, Yes/No confirmation options

natalie-perlin committed:
add individual plots to ensemble members in grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16 test

natalie-perlin commented:
@mkavulich @MichaelLueken - 
All updated. I've added plotting fields for individual ensemble members to the test grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16

It is however stuck in the queue, and shown as "PENDING" for few hours:

[Natalie.Perlin@Hera:/scratch2/NCEPDEV/stmp1/Natalie.Perlin/SRW/ufs-srweather-app]$ squeue -u Natalie.Perlin
     JOBID PARTITION  NAME                     USER             STATE        TIME TIME_LIMIT NODES NODELIST(REASON)
  65257017 hera       make_grid                Natalie.Perlin   PENDING      0:00      20:00     1 (Priority)
  65257038 hera       run_MET_PcpCombine_obs_A Natalie.Perlin   PENDING      0:00      30:00     1 (Priority)
  65257039 hera       run_MET_PcpCombine_obs_A Natalie.Perlin   PENDING      0:00      30:00     1 (Priority)
  65257040 hera       run_MET_PcpCombine_obs_A Natalie.Perlin   PENDING      0:00      30:00     1 (Priority)
  65257045 hera       run_MET_Pb2nc_obs        Natalie.Perlin   PENDING      0:00      30:00     1 (Priority)
  
  Not sure if there is an issue of running jobs on Hera. I could try using a different platform that may have a lighter compute load and may complete test faster... let me know!
  
  
Attached is a snapshot of the built documentation showing the update made for devclean.sh script.
![devclean_doc_updates](https://github.com/user-attachments/assets/6325378c-b246-4eda-98e0-3a84b59a9382)

natalie-perlin commented:
@MichaelLueken - the test grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16 with plots for each ensemble has completed successfully on Hera, in /scratch2/NCEPDEV/stmp1/Natalie.Perlin/SRW/expt_dirs/grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16

MichaelLueken reviewed (approved):
@natalie-perlin -

Thanks for making the requested changes!  I have successfully run the fundamental WE2E tests as well:
```
----------------------------------------------------------------------------------------------------
Experiment name                                                  | Status    | Core hours used
----------------------------------------------------------------------------------------------------
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta_2  COMPLETE              11.70
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2_20240  COMPLETE               8.55
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v17_p8_plot  COMPLETE              26.05
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR_2024081  COMPLETE              46.78
grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_WoFS_v0_20240819195  COMPLETE              27.05
grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16_2024081919530  COMPLETE              48.16
----------------------------------------------------------------------------------------------------
Total                                                              COMPLETE             168.29
```
and the plots for the ensemble test successfully completed as well:
```
plot_allvars_mem001_f000_202105121200     SUCCEEDED         205.0           1.37
plot_allvars_mem001_f001_202105121200     SUCCEEDED         204.0           1.36
plot_allvars_mem001_f002_202105121200     SUCCEEDED         204.0           1.36
plot_allvars_mem001_f003_202105121200     SUCCEEDED         205.0           1.37
plot_allvars_mem001_f004_202105121200     SUCCEEDED         204.0           1.36
plot_allvars_mem001_f005_202105121200     SUCCEEDED         205.0           1.37
plot_allvars_mem001_f006_202105121200     SUCCEEDED         205.0           1.37
plot_allvars_mem002_f000_202105121200     SUCCEEDED         205.0           1.37
plot_allvars_mem002_f001_202105121200     SUCCEEDED         204.0           1.36
plot_allvars_mem002_f002_202105121200     SUCCEEDED         204.0           1.36
plot_allvars_mem002_f003_202105121200     SUCCEEDED         204.0           1.36
plot_allvars_mem002_f004_202105121200     SUCCEEDED         206.0           1.37
plot_allvars_mem002_f005_202105121200     SUCCEEDED         205.0           1.37
plot_allvars_mem002_f006_202105121200     SUCCEEDED         205.0           1.37
```
It would be a good idea to modify the description for the `config.grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16.yaml` configuration to include that it also tests the ability to produce graphics for each ensemble member.

Since the fundamental WE2E tests are all running successfully (with the addition of adding plotting to the ensemble test in the fundamental suite), I will re-approve this work now.

natalie-perlin commented:
> @natalie-perlin -
> 
> Thanks for making the requested changes! I have successfully run the fundamental WE2E tests as well:
> 
> ```
> 
> It would be a good idea to modify the description for the `config.grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16.yaml` configuration to include that it also tests the ability to produce graphics for each ensemble member.
> 
> Since the fundamental WE2E tests are all running successfully (with the addition of adding plotting to the ensemble test in the fundamental suite), I will re-approve this work now.

@MichaelLueken - where the description of the test config.grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16.yaml
could be changed? Is it in the documentation or any comments in the code?

MichaelLueken commented:
> 
> @MichaelLueken - where the description of the test config.grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16.yaml could be changed? Is it in the documentation or any comments in the code?

@natalie-perlin - If you go directly to `tests/WE2E/test_configs/grids_extrn_mdls_suites_community/config.grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16.yaml`, the description is at the top of the file:
```
metadata:
  description: |-
    This test is to ensure that the workflow running in community mode
    completes successfully on the RRFS_CONUS_25km grid using the GFS_v16
    physics suite with ICs and LBCs derived from the NAM.
    This test also runs with two ensemble members, and ensures the MET
    ensemble-specific tasks run successfully.
```
It would be a good idea to include that either plots are produced or graphics are created for each ensemble member in the description section above.

natalie-perlin committed:
Update test description for config.grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16.yaml

natalie-perlin commented:
@MichaelLueken @mkavulich - updated the test description to the following:
```
This test also runs with two ensemble members, runs ploting tasks for each
    ensemble member, and ensures the MET ensemble-specific tasks run successfully.
 ```

mkavulich reviewed (approved):
@natalie-perlin Thanks for the productive discussion and back-and-forth, I'm glad we were able to reach a compromise

MichaelLueken commented:
The Jenkins tests successfully passed on all machines, but the test stage was ultimately terminated for running past the 8 hour limit on Jet.

The WE2E coverage tests were manually ran on Jet and all tests successfully passed:
```
----------------------------------------------------------------------------------------------------
Experiment name                                                  | Status    | Core hours used 
----------------------------------------------------------------------------------------------------
community_20240822191353                                           COMPLETE              22.88
custom_ESGgrid_20240822191354                                      COMPLETE              29.55
custom_ESGgrid_Great_Lakes_snow_8km_20240822191355                 COMPLETE              26.54
custom_GFDLgrid_20240822191357                                     COMPLETE              14.60
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2021032018_202408  COMPLETE              14.85
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_netcdf_2022060112_48h_20  COMPLETE             109.96
get_from_HPSS_ics_RAP_lbcs_RAP_20240822191400                      COMPLETE              20.22
grid_RRFS_AK_3km_ics_FV3GFS_lbcs_FV3GFS_suite_HRRR_20240822191400  COMPLETE             638.33
grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16_plot_20  COMPLETE              84.71
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2_20240  COMPLETE               8.62
grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta_2024  COMPLETE             962.58
----------------------------------------------------------------------------------------------------
Total                                                              COMPLETE            1932.84
```
Moving forward with merging this PR now.
